File: rust/lance-namespace/src/namespace.rs
//! Lance Namespace base interface and implementations.

use async_trait::async_trait;
use bytes::Bytes;
use std::collections::HashMap;
use thiserror::Error;

use lance_namespace_reqwest_client::models::{
    AlterTransactionRequest, AlterTransactionResponse, CountTableRowsRequest,
    CreateEmptyTableRequest, CreateEmptyTableResponse, CreateNamespaceRequest,
    CreateNamespaceResponse, CreateTableIndexRequest, CreateTableIndexResponse, CreateTableRequest,
    CreateTableResponse, DeleteFromTableRequest, DeleteFromTableResponse, DeregisterTableRequest,
    DeregisterTableResponse, DescribeNamespaceRequest, DescribeNamespaceResponse,
    DescribeTableIndexStatsRequest, DescribeTableIndexStatsResponse, DescribeTableRequest,
    DescribeTableResponse, DescribeTransactionRequest, DescribeTransactionResponse,
    DropNamespaceRequest, DropNamespaceResponse, DropTableRequest, DropTableResponse,
    InsertIntoTableRequest, InsertIntoTableResponse, ListNamespacesRequest, ListNamespacesResponse,
    ListTableIndicesRequest, ListTableIndicesResponse, ListTablesRequest, ListTablesResponse,
    MergeInsertIntoTableRequest, MergeInsertIntoTableResponse, NamespaceExistsRequest,
    QueryTableRequest, RegisterTableRequest, RegisterTableResponse, TableExistsRequest,
    UpdateTableRequest, UpdateTableResponse,
};

/// Error type for namespace operations
#[derive(Debug, Error)]
pub enum NamespaceError {
    #[error("Operation not supported: {0}")]
    NotSupported(String),

    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),

    #[error("Other error: {0}")]
    Other(String),
}

/// Result type for namespace operations
pub type Result<T> = std::result::Result<T, NamespaceError>;

/// Base trait for Lance Namespace implementations.
///
/// This trait defines the interface that all Lance namespace implementations
/// must provide. Each method corresponds to a specific operation on namespaces
/// or tables.
#[async_trait]
pub trait LanceNamespace: Send + Sync {
    /// List namespaces.
    async fn list_namespaces(
        &self,
        _request: ListNamespacesRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<ListNamespacesResponse> {
        Err(NamespaceError::NotSupported("list_namespaces".to_string()))
    }

    /// Describe a namespace.
    async fn describe_namespace(
        &self,
        _request: DescribeNamespaceRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<DescribeNamespaceResponse> {
        Err(NamespaceError::NotSupported(
            "describe_namespace".to_string(),
        ))
    }

    /// Create a new namespace.
    async fn create_namespace(
        &self,
        _request: CreateNamespaceRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<CreateNamespaceResponse> {
        Err(NamespaceError::NotSupported("create_namespace".to_string()))
    }

    /// Drop a namespace.
    async fn drop_namespace(
        &self,
        _request: DropNamespaceRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<DropNamespaceResponse> {
        Err(NamespaceError::NotSupported("drop_namespace".to_string()))
    }

    /// Check if a namespace exists.
    async fn namespace_exists(
        &self,
        _request: NamespaceExistsRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<()> {
        Err(NamespaceError::NotSupported("namespace_exists".to_string()))
    }

    /// List tables in a namespace.
    async fn list_tables(
        &self,
        _request: ListTablesRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<ListTablesResponse> {
        Err(NamespaceError::NotSupported("list_tables".to_string()))
    }

    /// Describe a table.
    async fn describe_table(
        &self,
        _request: DescribeTableRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<DescribeTableResponse> {
        Err(NamespaceError::NotSupported("describe_table".to_string()))
    }

    /// Register a table.
    async fn register_table(
        &self,
        _request: RegisterTableRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<RegisterTableResponse> {
        Err(NamespaceError::NotSupported("register_table".to_string()))
    }

    /// Check if a table exists.
    async fn table_exists(
        &self,
        _request: TableExistsRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<()> {
        Err(NamespaceError::NotSupported("table_exists".to_string()))
    }

    /// Drop a table.
    async fn drop_table(
        &self,
        _request: DropTableRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<DropTableResponse> {
        Err(NamespaceError::NotSupported("drop_table".to_string()))
    }

    /// Deregister a table.
    async fn deregister_table(
        &self,
        _request: DeregisterTableRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<DeregisterTableResponse> {
        Err(NamespaceError::NotSupported("deregister_table".to_string()))
    }

    /// Count rows in a table.
    async fn count_table_rows(
        &self,
        _request: CountTableRowsRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<i64> {
        Err(NamespaceError::NotSupported("count_table_rows".to_string()))
    }

    /// Create a new table with data from Arrow IPC stream.
    async fn create_table(
        &self,
        _request: CreateTableRequest,
        _request_data: Bytes,
        _config: Option<HashMap<String, String>>,
    ) -> Result<CreateTableResponse> {
        Err(NamespaceError::NotSupported("create_table".to_string()))
    }

    /// Create an empty table (metadata only operation).
    async fn create_empty_table(
        &self,
        _request: CreateEmptyTableRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<CreateEmptyTableResponse> {
        Err(NamespaceError::NotSupported(
            "create_empty_table".to_string(),
        ))
    }

    /// Insert data into a table.
    async fn insert_into_table(
        &self,
        _request: InsertIntoTableRequest,
        _request_data: Bytes,
        _config: Option<HashMap<String, String>>,
    ) -> Result<InsertIntoTableResponse> {
        Err(NamespaceError::NotSupported(
            "insert_into_table".to_string(),
        ))
    }

    /// Merge insert data into a table.
    async fn merge_insert_into_table(
        &self,
        _request: MergeInsertIntoTableRequest,
        _request_data: Bytes,
        _config: Option<HashMap<String, String>>,
    ) -> Result<MergeInsertIntoTableResponse> {
        Err(NamespaceError::NotSupported(
            "merge_insert_into_table".to_string(),
        ))
    }

    /// Update a table.
    async fn update_table(
        &self,
        _request: UpdateTableRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<UpdateTableResponse> {
        Err(NamespaceError::NotSupported("update_table".to_string()))
    }

    /// Delete from a table.
    async fn delete_from_table(
        &self,
        _request: DeleteFromTableRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<DeleteFromTableResponse> {
        Err(NamespaceError::NotSupported(
            "delete_from_table".to_string(),
        ))
    }

    /// Query a table.
    async fn query_table(
        &self,
        _request: QueryTableRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<Bytes> {
        Err(NamespaceError::NotSupported("query_table".to_string()))
    }

    /// Create a table index.
    async fn create_table_index(
        &self,
        _request: CreateTableIndexRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<CreateTableIndexResponse> {
        Err(NamespaceError::NotSupported(
            "create_table_index".to_string(),
        ))
    }

    /// List table indices.
    async fn list_table_indices(
        &self,
        _request: ListTableIndicesRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<ListTableIndicesResponse> {
        Err(NamespaceError::NotSupported(
            "list_table_indices".to_string(),
        ))
    }

    /// Describe table index statistics.
    async fn describe_table_index_stats(
        &self,
        _request: DescribeTableIndexStatsRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<DescribeTableIndexStatsResponse> {
        Err(NamespaceError::NotSupported(
            "describe_table_index_stats".to_string(),
        ))
    }

    /// Describe a transaction.
    async fn describe_transaction(
        &self,
        _request: DescribeTransactionRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<DescribeTransactionResponse> {
        Err(NamespaceError::NotSupported(
            "describe_transaction".to_string(),
        ))
    }

    /// Alter a transaction.
    async fn alter_transaction(
        &self,
        _request: AlterTransactionRequest,
        _config: Option<HashMap<String, String>>,
    ) -> Result<AlterTransactionResponse> {
        Err(NamespaceError::NotSupported(
            "alter_transaction".to_string(),
        ))
    }
}

File: typescript/lance-namespace/src/namespace.ts
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type * as Models from "@lance/namespace-fetch-client";

import { UnsupportedOperationError } from "./errors";

export abstract class LanceNamespace {
  public abstract namespaceId(): string;

  protected unsupported(name: string): never {
    throw new UnsupportedOperationError(`Not supported: ${name}`);
  }

  public async listNamespaces(
    _request: Models.ListNamespacesRequest,
  ): Promise<Models.ListNamespacesResponse> {
    return this.unsupported("listNamespaces");
  }

  public async describeNamespace(
    _request: Models.DescribeNamespaceRequest,
  ): Promise<Models.DescribeNamespaceResponse> {
    return this.unsupported("describeNamespace");
  }

  public async createNamespace(
    _request: Models.CreateNamespaceRequest,
  ): Promise<Models.CreateNamespaceResponse> {
    return this.unsupported("createNamespace");
  }

  public async dropNamespace(
    _request: Models.DropNamespaceRequest,
  ): Promise<Models.DropNamespaceResponse> {
    return this.unsupported("dropNamespace");
  }

  public async namespaceExists(
    _request: Models.NamespaceExistsRequest,
  ): Promise<void> {
    return this.unsupported("namespaceExists");
  }

  public async listTables(
    _request: Models.ListTablesRequest,
  ): Promise<Models.ListTablesResponse> {
    return this.unsupported("listTables");
  }

  public async listAllTables(
    _request: Models.ListTablesRequest,
  ): Promise<Models.ListTablesResponse> {
    return this.unsupported("listAllTables");
  }

  public async describeTable(
    _request: Models.DescribeTableRequest,
  ): Promise<Models.DescribeTableResponse> {
    return this.unsupported("describeTable");
  }

  public async registerTable(
    _request: Models.RegisterTableRequest,
  ): Promise<Models.RegisterTableResponse> {
    return this.unsupported("registerTable");
  }

  public async tableExists(_request: Models.TableExistsRequest): Promise<void> {
    return this.unsupported("tableExists");
  }

  public async dropTable(
    _request: Models.DropTableRequest,
  ): Promise<Models.DropTableResponse> {
    return this.unsupported("dropTable");
  }

  public async deregisterTable(
    _request: Models.DeregisterTableRequest,
  ): Promise<Models.DeregisterTableResponse> {
    return this.unsupported("deregisterTable");
  }

  public async createTable(
    _request: Models.CreateTableRequest,
    _requestData: Uint8Array,
  ): Promise<Models.CreateTableResponse> {
    return this.unsupported("createTable");
  }

  public async declareTable(
    _request: Models.DeclareTableRequest,
  ): Promise<Models.DeclareTableResponse> {
    return this.unsupported("declareTable");
  }

  public async createEmptyTable(
    _request: Models.CreateEmptyTableRequest,
  ): Promise<Models.CreateEmptyTableResponse> {
    return this.unsupported("createEmptyTable");
  }

  public async insertIntoTable(
    _request: Models.InsertIntoTableRequest,
    _requestData: Uint8Array,
  ): Promise<Models.InsertIntoTableResponse> {
    return this.unsupported("insertIntoTable");
  }

  public async mergeInsertIntoTable(
    _request: Models.MergeInsertIntoTableRequest,
    _requestData: Uint8Array,
  ): Promise<Models.MergeInsertIntoTableResponse> {
    return this.unsupported("mergeInsertIntoTable");
  }

  public async updateTable(
    _request: Models.UpdateTableRequest,
  ): Promise<Models.UpdateTableResponse> {
    return this.unsupported("updateTable");
  }

  public async deleteFromTable(
    _request: Models.DeleteFromTableRequest,
  ): Promise<Models.DeleteFromTableResponse> {
    return this.unsupported("deleteFromTable");
  }

  public async queryTable(
    _request: Models.QueryTableRequest,
  ): Promise<Uint8Array> {
    return this.unsupported("queryTable");
  }

  public async countTableRows(
    _request: Models.CountTableRowsRequest,
  ): Promise<number> {
    return this.unsupported("countTableRows");
  }

  public async createTableIndex(
    _request: Models.CreateTableIndexRequest,
  ): Promise<Models.CreateTableIndexResponse> {
    return this.unsupported("createTableIndex");
  }

  public async createTableScalarIndex(
    _request: Models.CreateTableIndexRequest,
  ): Promise<Models.CreateTableScalarIndexResponse> {
    return this.unsupported("createTableScalarIndex");
  }

  public async listTableIndices(
    _request: Models.ListTableIndicesRequest,
  ): Promise<Models.ListTableIndicesResponse> {
    return this.unsupported("listTableIndices");
  }

  public async describeTableIndexStats(
    _request: Models.DescribeTableIndexStatsRequest,
  ): Promise<Models.DescribeTableIndexStatsResponse> {
    return this.unsupported("describeTableIndexStats");
  }

  public async dropTableIndex(
    _request: Models.DropTableIndexRequest,
  ): Promise<Models.DropTableIndexResponse> {
    return this.unsupported("dropTableIndex");
  }

  public async restoreTable(
    _request: Models.RestoreTableRequest,
  ): Promise<Models.RestoreTableResponse> {
    return this.unsupported("restoreTable");
  }

  public async renameTable(
    _request: Models.RenameTableRequest,
  ): Promise<Models.RenameTableResponse> {
    return this.unsupported("renameTable");
  }

  public async listTableVersions(
    _request: Models.ListTableVersionsRequest,
  ): Promise<Models.ListTableVersionsResponse> {
    return this.unsupported("listTableVersions");
  }

  public async createTableVersion(
    _request: Models.CreateTableVersionRequest,
  ): Promise<Models.CreateTableVersionResponse> {
    return this.unsupported("createTableVersion");
  }

  public async describeTableVersion(
    _request: Models.DescribeTableVersionRequest,
  ): Promise<Models.DescribeTableVersionResponse> {
    return this.unsupported("describeTableVersion");
  }

  public async batchDeleteTableVersions(
    _request: Models.BatchDeleteTableVersionsRequest,
  ): Promise<Models.BatchDeleteTableVersionsResponse> {
    return this.unsupported("batchDeleteTableVersions");
  }

  public async batchCreateTableVersions(
    _request: Models.BatchCreateTableVersionsRequest,
  ): Promise<Models.BatchCreateTableVersionsResponse> {
    return this.unsupported("batchCreateTableVersions");
  }

  public async updateTableSchemaMetadata(
    _request: Models.UpdateTableSchemaMetadataRequest,
  ): Promise<Models.UpdateTableSchemaMetadataResponse> {
    return this.unsupported("updateTableSchemaMetadata");
  }

  public async getTableStats(
    _request: Models.GetTableStatsRequest,
  ): Promise<Models.GetTableStatsResponse> {
    return this.unsupported("getTableStats");
  }

  public async explainTableQueryPlan(
    _request: Models.ExplainTableQueryPlanRequest,
  ): Promise<string> {
    return this.unsupported("explainTableQueryPlan");
  }

  public async analyzeTableQueryPlan(
    _request: Models.AnalyzeTableQueryPlanRequest,
  ): Promise<string> {
    return this.unsupported("analyzeTableQueryPlan");
  }

  public async alterTableAddColumns(
    _request: Models.AlterTableAddColumnsRequest,
  ): Promise<Models.AlterTableAddColumnsResponse> {
    return this.unsupported("alterTableAddColumns");
  }

  public async alterTableAlterColumns(
    _request: Models.AlterTableAlterColumnsRequest,
  ): Promise<Models.AlterTableAlterColumnsResponse> {
    return this.unsupported("alterTableAlterColumns");
  }

  public async alterTableDropColumns(
    _request: Models.AlterTableDropColumnsRequest,
  ): Promise<Models.AlterTableDropColumnsResponse> {
    return this.unsupported("alterTableDropColumns");
  }

  public async listTableTags(
    _request: Models.ListTableTagsRequest,
  ): Promise<Models.ListTableTagsResponse> {
    return this.unsupported("listTableTags");
  }

  public async getTableTagVersion(
    _request: Models.GetTableTagVersionRequest,
  ): Promise<Models.GetTableTagVersionResponse> {
    return this.unsupported("getTableTagVersion");
  }

  public async createTableTag(
    _request: Models.CreateTableTagRequest,
  ): Promise<Models.CreateTableTagResponse> {
    return this.unsupported("createTableTag");
  }

  public async deleteTableTag(
    _request: Models.DeleteTableTagRequest,
  ): Promise<Models.DeleteTableTagResponse> {
    return this.unsupported("deleteTableTag");
  }

  public async updateTableTag(
    _request: Models.UpdateTableTagRequest,
  ): Promise<Models.UpdateTableTagResponse> {
    return this.unsupported("updateTableTag");
  }

  public async describeTransaction(
    _request: Models.DescribeTransactionRequest,
  ): Promise<Models.DescribeTransactionResponse> {
    return this.unsupported("describeTransaction");
  }

  public async alterTransaction(
    _request: Models.AlterTransactionRequest,
  ): Promise<Models.AlterTransactionResponse> {
    return this.unsupported("alterTransaction");
  }
}
